tests/integration/utils/network.py

```diff
@@ -3,81 +3,72 @@
+        self.__bound: Set[int] = set()
 
     @contextmanager
-    def allocate_port(self) -> int:
-        """Find a random port in the range `PortRangeInclusive`.
+    def allocate_port(self) -> ContextManager[int]:
+        """Allocates the next free port in this range for binding until exit of the context.
 
-        Note:
-            This function is *not* aware of the ports currently open in the system,
-            the blacklist only prevents two services of the same type to randomly
-            get the same ports for *a single test run*.
+        The yielded port is guaranteed to be free and unusable by any other part of this process that uses this port
+        range to allocate ports. However, other processes may bind to the port during the time between the check if the
+        port is free and the caller actually binding to it. This time window increases the more ports are being
+        allocated. There is nothing that can be done about this, other than using software that can find a free port on
+        its own.
 
-            Because of that, the port range should be chosen such that there is no
-            system service in the range. Also note that running two sessions of the
-            tests with the same range is not supported and will lead to flakiness.
+        :raises RuntimeException: if no ports are free in this range.
         """
-        if len(self._maybe_available) == 0:
-            raise RuntimeError(f"No free ports available. start: {self.start} end: {self.end}")
-
-        filtered_ports = ((pos, port) for pos, port in enumerate(self._maybe_available) if not is_time_wait(port))
-
-        try:
-            pos, port = next(filtered_ports)
-        except StopIteration as e:
-            raise RuntimeError(
-                f"No free ports available. start: {self.start} end: {self.end} time_wait: {self._maybe_available}"
-            ) from e
-
-        self._maybe_available.pop(pos)
-        yield port
-
-        # Append the port at the end, this is a hack to give extra time for a TIME_WAIT socket to
-        # close, but it is not sufficient.
-        self._maybe_available.append(port)
-
-
-def port_is_listening(hostname: str, port: int, ipv6: bool) -> bool:
-    if ipv6:
-        s = socket.socket(socket.AF_INET6, socket.SOCK_STREAM, 0)
-    else:
-        s = socket.socket()
-    s.settimeout(0.5)
-    try:
-        s.connect((hostname, port))
-        s.close()
-        return True
-    except socket.error:
-        return False
+        sock = socket.socket()
+        for port in range(self.__start, self.__end + 1):
+            # We gave this port to a previous caller, and thus cannot use it, even if it is free. This is always a small
+            # collection, as we are not expecting tests to bind to hundreds of ports.
+            if port not in self.__bound:
+                self.__bound.add(port)
+                try:
+                    sock.bind(("127.0.0.1", port))
```

Comment:
What about something like this, without looping and would also remove the need for min and max port?
```suggestion
import closing
with closing(socket.socket.socket(socket.AF_INET, socket.SOCK_STREAM)) as sock:
    sock.bind(("127.0.0.1", 0))
    s.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
    port = sock.getsockname()[1]
yield port
```

Reply:
We could get rid of the entire port range machinery and instead just search the entire port space, yes. I had a look at the whole range logic, but decided to not touch too much. But, we can do that. Less code is always better. 😊

Reply:
Thinking of it, this actually is not going to work. From what I understand the tests are executed in parallel with different processes. This means that one process might find a free port, hand it out, and another process finds the same port to be free since it was not bound to in the other process just yet. Now we have two components that want to bind to the same port. I think the logic as it is is necessary.